### ISSUE TITLE
Unresolved specs during Gem::Specification.reset (rake)

### ISSUE THREAD

**Issue by dialex** [inspec/inspec#2105]
### Description

I get a warning every time I run inspec:

```
WARN: Unresolved specs during Gem::Specification.reset:
      rake (>= 0)
WARN: Clearing out unresolved specs.
Please report a bug if this causes problems.
```

### InSpec and Platform Version

Version: 1.34.1
Operating System: Ubuntu (actually a Docker image from [Microsoft](https://hub.docker.com/r/microsoft/vsts-agent/))

### Replication Case

1. Pull the docker image.
2. Install ruby: `RUN apt-get update && apt-get install -y ruby-full`
3. Install bundler and rake: `RUN gem install -N rake bundler`
4. Install inspec `RUN gem install inspec`
5. Run `$ inspec`.

### Possible Solutions

- After steps 1 to 4, I did `gem list rake` and I already get:

```
*** LOCAL GEMS ***
rake (12.0.0, 10.5.0)
```
- ...thus it's not my Gemfile or the inspec installation that is adding the "additional" version of rake.
- I tried all solutions at [SO](https://stackoverflow.com/questions/17936340/unresolved-specs-during-gemspecification-reset) and none worked.
  - `gem cleanup` did nothing.
  - `bundle clean --force` did nothing.
  - `gem uninstall rake 10.5.0` actually removed rake 12.0.0 and turned the command unrecognised.

### Stacktrace

(tell me what is useful to you)

**Comment by cheeseplus** [comment]
You mention a Gemfile but it doesn't appear you're using bundler anywhere in the above steps. This is a classic Rubygems warning that to resolve you'll need to make sure that InSpec is either isolated in a bundle install and executed via `bundle exec` or install InSpec via it's own installer/ChefDK which both effectively bundle ruby and the gemssets so you don't have to do so.

Even then it's possible to still get this warning but it's also just a warning and should doesn't break anything.

**Comment by adamleff** [comment]
Hi @dialex, thanks for the issue!

This issue isn't specific to InSpec but rather the Ruby environment you're installing in your container. The `ruby-full` APT package has some weirdnesses with it that include a non-default directory in the GEM_PATH that the `gem` command won't look into unless you specifically tell it to.

This error specifically means that it couldn't load `rake` (even though InSpec itself doesn't depend on it, some of the gems InSpec uses do use Rake to build native extensions)... in this case, it couldn't load it because there are two versions installed (12.0.0 and 10.5.0).

I could easily reproduce your issue with this Dockerfile:

```
FROM ubuntu

RUN apt-get update && apt-get install -y ruby-full gcc make
RUN gem install rake inspec
```

Now, as @cheeseplus mentioned, it's not actually a big deal - it's just a warning. However, if you want to get rid of the warning, you need to uninstall one of the versions of rake. I recommend uninstalling the old one, and it's unfortunately in that non-standard path I mentioned.

With this Dockerfile, I no longer have the issue you reported:

```
FROM ubuntu

RUN apt-get update && apt-get install -y ruby-full gcc make
RUN gem install rake inspec
RUN gem uninstall -i /usr/share/rubygems-integration/all rake 10.5.0
```

I hope that helps. Also, please note that we do publish each of our stable InSpec releases to `chef/inspec` which may just be able to use as a layer in your container.

Thanks!
~Adam

**Comment by dialex** [comment]
Thank you so much Adam! Even though it was not related with InSpec, you gave a detailed answer that solves my problem. Cheers mate 🙏

**Comment by b-dean** [comment]
I get the same error for this `Dockerfile`. 

```dockerfile
FROM centos:7
RUN yum install -y curl
RUN curl -L https://omnitruck.chef.io/install.sh | bash -s -- -P inspec
```
(note that this is in the context of a larger Dockerfile with other stuff, but this smaller example also has the problem).

Then when I run 

```
[root@ded8fce0755d /]# inspec --version
WARN: Unresolved specs during Gem::Specification.reset:
      rake (>= 0)
WARN: Clearing out unresolved specs.
Please report a bug if this causes problems.
1.45.13
```

I'm not sure I understand the explanation @adamleff gave earlier. There shouldn't be any weird GEM_PATH stuff going on, I installed the omnibus package. I wouldn't think it'd have more than one rake either but it does:

```
[root@ded8fce0755d /]# /opt/inspec/embedded/bin/gem list rake 

*** LOCAL GEMS ***

rake (12.0.0, 10.5.0)
```

If I uninstall the older rake then I don't get the warning. But it still is concerning to me since it showed up in the omnibus package which means _something_ wanted it.

```
[root@ded8fce0755d /]# /opt/inspec/embedded/bin/gem uninstall rake -v 10.5.0
Successfully uninstalled rake-10.5.0
[root@ded8fce0755d /]# inspec --version
1.45.13
```

**Comment by twdamhore** [comment]
I hit the same error. Solution provided by @b-dean worked for me. 
``` /opt/inspec/embedded/bin/gem list | grep rake```
Make sure there is only one version of rake installed.

**Comment by adamleff** [comment]
If you all are getting this with our omnibus packages, that needs further investigation. I’ll reopen this issue.

**Comment by ryancurrah** [comment]
I'm getting this issue with the RPM provided. I do not have Ruby installed on my vagrantbox

```
[vagrant@rhel ~]$ /opt/inspec/embedded/bin/gem list | grep rake
rake (12.0.0, 10.5.0)
```

**Comment by adamleff** [comment]
I have tracked down this issue and it is fixed in #2374 and should be fixed in the next stable release of InSpec. Thanks for your patience.
